oauth-common/src/main/java/io/strimzi/kafka/oauth/common/OAuthAuthenticator.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright 2017-2019, Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.kafka.oauth.common;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.net.ssl.HostnameVerifier;
+import javax.net.ssl.SSLSocketFactory;
+import java.io.IOException;
+import java.io.UnsupportedEncodingException;
+import java.net.URI;
+import java.net.URLEncoder;
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+
+import static io.strimzi.kafka.oauth.common.LogUtil.mask;
+import static io.strimzi.kafka.oauth.common.TokenIntrospection.introspectAccessToken;
+
+public class OAuthAuthenticator {
+
+    private static final Logger log = LoggerFactory.getLogger(OAuthAuthenticator.class);
+
+    public static TokenInfo loginWithAccessToken(String token) {
+        if (log.isDebugEnabled()) {
+            log.debug("loginWithAccessToken() - pass-through access_token: {}", mask(token));
+        }
+        // try introspect token
+        try {
+            return introspectAccessToken(token);
+        } catch (Exception e) {
+            log.debug("[IGNORED] Could not parse token as JWT access token. Could not extract scope, subject, and expiry.", e);
+        }
+        return new TokenInfo(token, "undefined", "undefined", System.currentTimeMillis(), System.currentTimeMillis() + 365 * 24 * 3600000);
+    }
+
+    public static TokenInfo loginWithClientSecret(URI tokenEndpointUrl, SSLSocketFactory socketFactory, HostnameVerifier hostnameVerifier, String clientId, String clientSecret) throws IOException {
+        if (log.isDebugEnabled()) {
+            log.debug("loginWithClientSecret() - tokenEndpointUrl: {}, clientId: {}, clientSecret: {}",
+                    tokenEndpointUrl, clientId, mask(clientSecret));
+        }
+        String authorization = "Basic " + base64encode(clientId + ':' + clientSecret);
+
+        StringBuilder body = new StringBuilder("grant_type=client_credentials");
+
+        return post(tokenEndpointUrl, socketFactory, hostnameVerifier, authorization, body.toString());
+    }
+
+    public static TokenInfo loginWithRefreshToken(URI tokenEndpointUrl, SSLSocketFactory socketFactory, HostnameVerifier hostnameVerifier, String refreshToken, String clientId, String clientSecret) throws IOException {
+        if (log.isDebugEnabled()) {
+            log.debug("loginWithRefreshToken() - tokenEndpointUrl: {}, refreshToken: {}, clientId: {}, clientSecret: {}",
+                    tokenEndpointUrl, refreshToken, clientId, mask(clientSecret));
+        }
+
+        String authorization = clientSecret != null ?
+                "Basic " + base64encode(clientId + ':' + clientSecret) :
+                null;
+
+        StringBuilder body = new StringBuilder("grant_type=refresh_token")
+                .append("&refresh_token=").append(refreshToken)
```

Comment:
`refreshToken` doesn't need url encoding?

Reply:
Well, if it is a JWT refresh token then it won't contain any characters that would require it to be url encoded. But if it is not JWT then I guess it could. So in first case url encoding would be a noop, but in second case it would be required. I'll add it.